src/Lorentz/Contracts/BaseDAO/Proposal.hs

```diff
@@ -604,63 +577,104 @@ unfreezeVoterToken = do
+  if #sQuorumThreshold <=. #proposalTotalVote then
+    push True
+  else
+    push False
+
+-- Delete a finished proposal from 'sProposalKeyListSortByDate'
+deleteProposalKeyFromMap
+  :: forall store ce pm s.
+     (NiceParameter pm, StorageC store ce pm)
+  => Proposal pm : store : s
+  :-> store : s
+deleteProposalKeyFromMap = do
+  toField #pStartDate
+  dip $ stGetField #sProposalKeyListSortByDate
+  dip $ none
+  update
+  stSetField #sProposalKeyListSortByDate
+
 handleProposalIsOver
   :: forall store ce pm s.
      (NiceParameter pm, StorageC store ce pm, HasFuncContext s store (AllFuncs store ce pm))
   => Config ce pm
-  -> ((Bool, ProposalKey pm), Proposal pm) : store : [Operation] : s
+  -> (Timestamp, ProposalKey pm) : store : [Operation] : s
   :-> store : [Operation] : s
 handleProposalIsOver config@Config{..} = do
-  unpair; unpair
-
+  cdr; dupTop2; checkIfProposalExist
+  stackType @(Proposal pm : ProposalKey pm : store : [Operation] : s)
+  dip swap
+  dupTop2; isVotingPeriodOver
   if Holds then do
-    -- start unfreezing process
-    swap; dip swap
-    stackType @(Proposal pm : store : ProposalKey pm : [Operation] : s)
-
-    getField #pUpvotes
-    dip $ getField #pDownvotes
-    add; toNamed #proposalTotalVote
-
-    duupX @3
-    stToField #sQuorumThreshold; toNamed #sQuorumThreshold
-
-    if #sQuorumThreshold <=. #proposalTotalVote then do
+    dupTop2; doTotalVoteMeetQuorumThreshold
+    if Holds then do
       getFieldNamed #pUpvotes
       dip $ getFieldNamed #pDownvotes
       if #pUpvotes >. #pDownvotes then do
+        stackType @(Proposal pm : store : ProposalKey pm : [Operation] : s)
         push (#isAccepted .! True)
         callCachedFunc $ unfreezeProposerToken config
         unfreezeVoterToken
+
+        dup; dip swap
         cDecisionLambda
-        stackType @([Operation] : store : ProposalKey pm : [Operation] : s)
-        dig @3
-        iter cons
-        swap; dip swap
+
+        stackType @([Operation] : store : Proposal pm : ProposalKey pm : [Operation] : s)
+        dig @4
+        iter cons; swap
+        stackType @(store : [Operation] : Proposal pm : ProposalKey pm : s)
+        dig @2;
+        deleteProposalKeyFromMap
+        dip $ dip drop
       else do
         -- Reject proposal if (upvote <= downvote)
+        stackType @(Proposal pm : store : ProposalKey pm : [Operation] : s)
         push (#isAccepted .! False)
         callCachedFunc $ unfreezeProposerToken config
         unfreezeVoterToken
-        drop
-
+        deleteProposalKeyFromMap
+        dip drop
     else do
-      -- Reject proposal regardless of upvotes
+      -- Reject proposal regardless of upvote
+      stackType @(Proposal pm : store : ProposalKey pm : [Operation] : s)
       push (#isAccepted .! False)
       callCachedFunc $ unfreezeProposerToken config
       unfreezeVoterToken
-      drop
-
-    dip $ drop @(ProposalKey pm)
-
+      deleteProposalKeyFromMap
+      dip drop
```

Comment:
Maybe we could also optimize here?: these two `else` branches seem to be identical but pretty big. Maybe we could construct a complex condition using `and`/`or` and have only one `if ... then ... else`? That should save a really lot of space.

Though a sad truth is that Lorentz does not have operators like `&&` and adding them there will take too long (no one severely used this functionality until baseDAO, I added [one](https://gitlab.com/morley-framework/morley/-/issues/505) and [two](https://gitlab.com/morley-framework/morley/-/issues/506) now), but probably we could add a helper that handles our specific condition.